README.md
Web Serial API binding for [Node.js serialport](https://serialport.io/)

```ts
import SerialPort from '@serialport/stream';
import WSABinding from 'serialport-binding-webserialapi';

SerialPort.Binding = WSABinding;

SerialPort.list().then(portsList => {
    console.log(JSON.stringify(portsList));
}, err => {
    console.error(JSON.stringify(err));
});
```

## Installation

This is a [Node.js](https://nodejs.org/en/) module available through the [npm registry](https://www.npmjs.com/).
Before installing, [download and install Node.js](https://nodejs.org/en/download/).
If this is a brand new project, make sure to create a `package.json` first with the [`npm init` command](https://docs.npmjs.com/creating-a-package-json-file).
Installation is done using the [`npm install` command](https://docs.npmjs.com/getting-started/installing-npm-packages-locally):

```bash
$ npm install serialport-binding-webserialapi
```

As this module provides [Web Serial API](https://wicg.github.io/serial/) bindings for [node.js serialport](https://serialport.io/), applications using it will require `@serialport/stream` module installed as well. This is provided for You by Node.js dependency mechanism.

## Browser support

On 29th Dec 2020, Web Serial API is supported in modern Chromium based browsers, that is Chrome, Edge, Opera and of course Chromium.

### Special activities needed to enable Web Serial API support

  * Chromium based browsers require enabling **Experimental Web Platform Features**
  	Chrome & Chromium: `chrome://flags/#enable-experimental-web-platform-features`
	Opera: `opera://flags/#enable-experimental-web-platform-features`
	Edge: `edge://flags/#enable-experimental-web-platform-features`
  * On Linux, snap based Chromium installation requires connecting Chromium snap to USB mapped serial ports
    `sudo snap connect chromium:raw-usb`

## Features

  * Enables to use Node.js serialport module based apps in browser almost directly
  * Supports base read/write operations and setting/getting serial port flags
  * Supports USB Vendor & Product ID based port selection limitation

## Docs

Usage of Web Serial API is straightforward, as it does not expose new functionality, merely binds the well documented Node Serialport to the Web Serial API available in browsers. See the example below for usage and referenced documentation.

  * [Web Serial API](https://wicg.github.io/serial/) - Draft Communit Group Report, Web Serial API reference
  * [Node Serialport](https://serialport.io/) - Documentation for the Serialport module
  
### Issuses & Limitations

  * `update` command is not implemented, as there is no support in Web Serial API for baud rate change for open port
  * `flush` is not implemented, no specific function in Web Serial API; need to check if possible thru the underlying stream objects
  * `drain` is not implemented, no specific function in Web Serial API; need to check if possible thru the underlying stream objects or `write` function `Promise`
  * as of 1.0.3 release date, getPorts() in Web Serial API as implemented in Chromium does not return list of serial ports allowed by user, this means at every start of application user will have to select serial port in a popup shown by the browser
  * tested only in Chromium 87.0.4280.88
  * Web Serial API specification is a living document and it's implementation is experimental; things might break any moment

## Example

  This is a very simple TypeScript example of a browser serial terminal application using jQuery, Serialport and Web Serial API

```ts
// very simple serial terminal
import $ from 'jquery';
import './vendor';
import SerialPort from '@serialport/stream';
import WSABinding from 'serialport-binding-webserialapi';
import stripAnsi from 'strip-ansi';

// terminal settings
// does not echo
var term_speed = 115200;

// serial port binding to Web Serial API
SerialPort.Binding = WSABinding;

// construct simple webpage
var myPort = undefined;
var actref = $(document.createElement("button"));
var twinref = $(document.createElement("textarea"));
var bref = $("body");
bref.html('');
bref.append(actref);
bref.append(twinref);
bref.children().css("display", "block").css("font-family", "monospace");
twinref.prop( "disabled", true );
twinref.attr("rows", "24");
twinref.attr("cols", "80");
actref.text('Start serial terminal');
var ta = $('textarea');
window.setInterval(() => { 
    var tal = ta.val().length;
    ta.scrollTop(ta[0].scrollHeight);
    ta[0].setSelectionRange(tal, tal);
}, 100);

// handle terminal startup - must be user initiated!
actref.click(() => {
    myPort = new SerialPort('wsa://default', {
        baudRate: term_speed,
        autoOpen: true
    });
    myPort.on('data', data => {
        for(var i = 0; i<data.length; i++) {
            if (data[i] == 8) { // simplistic
                ta.val(ta.val().slice(0,-1));
            }
        }
        var re = /[\0-\x1F\x7F-\x9F\xAD]/;
        ta.val(ta.val() + stripAnsi(data.toString()).replace(re,''));
    });
    myPort.on('open', () => {
        ta.prop( "disabled", false );
        ta.focus();
    })
})

// handle terminal text input
function twrite(c:string) {
    myPort.write(c, err => {
        if (err) {
            console.log('Error on write: ', err.message);                    
        }
    });
}
twinref.on('keydown', e => {
    var tval = undefined;
    if (e.which == 13) {
        tval = "\r\n";
    } else if (e.which == 8) {
        tval = "\b";
    } else if (e.which == 9) {
        tval = "\t";
    } else {
        return;
    }
    twrite(tval);
    e.preventDefault();  
});
twinref.on('keyup', e => {
    e.preventDefault();
});
twinref.on('keypress', e => {
    twrite(String.fromCharCode(e.which));
    e.preventDefault();
});
```

## People

The original author is [Marcin Galczynski](https://github.com/nathanjel/)

## License

  [MIT](LICENSE)

CHANGELOG.md
# Changelog

## [v1.0.3](https://github.com/nathanjel/serialport-binding-web-serial-api/releases/tag/v1.0.0)(2020-12-29)

- first release
- tested in Chromium on Linux
- `flush`, `drain` and baudrate `update` not supported

---


wsa-sp-binds.ts
import AbstractBinding from '@serialport/binding-abstract';
import { strict } from 'assert';
import SerialPort, { PortInfo } from 'serialport';

declare global {
  interface Navigator {
    readonly serial: unknown;
  }
}

const internalBufferSize: number = 16384;
const internalNavigatorSerial: any = navigator.serial;
const backendAvailable: boolean = navigator.serial !== undefined;
let internalBasePortsList: [...any] = [];

function internalLog(text: string) {
  // console.log("%c WSA Log: " + text, 'background: #222; color: #bada55');
}

function dec2hex16bitWithPad(i: number): string {
  return (i + 0x10000).toString(16).substr(-4).toUpperCase();
}

function mapBasePortToPath(sourceSerialPort: any): string {
  const sourceInfo: any = sourceSerialPort.getInfo();
  return 'wsa://' + dec2hex16bitWithPad(sourceInfo.usbVendorId) + '-' + dec2hex16bitWithPad(sourceInfo.usbProductId);
}

function mapBasePortToPortInfo(sourceSerialPort: any): PortInfo {
  const sourceInfo: any = sourceSerialPort.getInfo();
  return {
    path: mapBasePortToPath(sourceSerialPort),
    productId: sourceInfo.usbProductId,
    vendorId: sourceInfo.usbVendorId
  };
}

function updatePortsList(): Promise<void> {
  let tPromise = new Promise(resolve => {
    resolve(internalBasePortsList.length == 0 ? null : internalBasePortsList[0]);
  });
  if (internalBasePortsList.length == 0) {
    tPromise = internalNavigatorSerial.requestPort();
  }
  return tPromise.then(port => {
    internalBasePortsList = [port];
    return internalNavigatorSerial.getPorts();
  }).then(ports => {
    return new Promise(resolve => {
      if (ports.length > 0) {
        internalBasePortsList = ports;
      }
      resolve(null);
    });
  });
}

function mapOpenOptions(input: SerialPort.OpenOptions): any {
  return {
    baudRate: input.baudRate,
    dataBits: input.dataBits,
    stopBits: input.stopBits,
    parity: input.parity,
    bufferSize: internalBufferSize,
    flowControl: input.rtscts ? "hardware" : "none"
  }
}

export default class WSABinding extends AbstractBinding {
  private storedOpenOptions: SerialPort.OpenOptions = {
    baudRate : 115200,
    dataBits: 8,
    stopBits: 1,
    parity: 'none',
    rtscts: false
  };
  private boundedPort: any | undefined = undefined;
  private actualBaudRate: number = -1;
  private classReader:ReadableStreamDefaultReader | undefined = undefined;
  private pendingRead:any|undefined = undefined;

  public isOpen: boolean = false;

  static list(): Promise<SerialPort.PortInfo[]> {
    return new Promise((resolve, reject) => {
      if (backendAvailable) {
        updatePortsList().then(res => {
          resolve(internalBasePortsList.map(mapBasePortToPortInfo))
        });
      } else {
        reject();
      }
    });
  }

  constructor(opt: SerialPort.OpenOptions) {
    super(opt);
    Object.assign(this.storedOpenOptions, opt);
  }

  open(path: string, options: SerialPort.OpenOptions): Promise<void> {
    internalLog("Open called for " + path);
    this.classReader = undefined;
    return super.open(path, options).then(ok => {
      return updatePortsList().then(ports => {
        for(var port of internalBasePortsList) {
          const mappedName = mapBasePortToPath(port);
          internalLog("Checking " + mappedName);
          if (mappedName == path) {
            this.boundedPort = port;
            let localOptions = this.storedOpenOptions;
            Object.assign(localOptions, options);
            return this.boundedPort.open(mapOpenOptions(localOptions)).then(okVoid => {
              internalLog("Open successful");
              this.isOpen = true;
              return new Promise(resolve => {
                internalLog("Open resolved");
                resolve(null);
              });
            }, failAny => {
              this.isOpen = false;
              internalLog("Open failed");
              return new Promise(reject => {
                internalLog("Open rejected with " + failAny);
                reject(null);
              })
            });
          }
        }
      });
    });
  }

  closePromise(): Promise<void> {
    return this.boundedPort.close().then(res => {
      return new Promise(resolve => {
        this.isOpen = false;
        resolve(null);
      });
    });
  }

  close(): Promise<void> {
    internalLog("Close called");
    return super.close().then(ok => {
      if (this.classReader) {
        this.classReader.cancel().then(ok => {
          this.classReader.releaseLock();
          return this.closePromise();        
        });
      } else {
        return this.closePromise();
      }
    });
  }

  write(buffer: Buffer): Promise<void> {
    internalLog("Write called");
    const mWriter = this.boundedPort.writable.getWriter();
    return super.write(buffer).then(ok => {
      return mWriter.write(buffer).then(res => {
        return new Promise(resolve => {
          internalLog("Write system execution completed");
          mWriter.releaseLock();
          resolve(null);
        });
      });
    });
  }

  registerInternalRead(): Promise<ReadableStreamReadResult<any>> {
    let internalRead = this.classReader.read();
    this.pendingRead = internalRead;
    return internalRead;
  }

  clearInternalRead() {
    this.pendingRead = undefined;
  }

  read(buffer: Buffer, offset: number, length: number): Promise<{ bytesRead: number, buffer: Buffer }> {
    let internalOffset:number = offset;
    let internalBytesRead: number = 0;
    let callbackCounter: number = 0;
    let currentRef:WSABinding = this;

    internalLog("Read called, expecting maximum so many bytes: " + length);
    const basePromise:Promise<any> = super.read(buffer, offset, length);

    internalLog("Super promise requested");

    function callReadAndPromise(): Promise<{ bytesRead: number, buffer: Buffer }> {
      internalLog("Read system execution initiated " + callbackCounter);
      return currentRef.registerInternalRead().then(readStruct => {
        internalLog("Read system execution completed " + callbackCounter);
        currentRef.clearInternalRead();
        return internalPromise(readStruct);
      });
    }

    function internalPromise(readerObject: { value?: any, done: boolean }) : Promise<{ bytesRead: number, buffer: Buffer }> {
      return new Promise<{ bytesRead: number, buffer: Buffer }>((resolve, reject) => {        
        internalLog("Updating result buffer " + callbackCounter);
        const done = readerObject.done;
        if (readerObject.value) {
        // we ignore the "done" flag, as it seems never to
        // reach the point of being set, even when there is
        // no more data to read, the Web Serial Api will just block
        // if (!done) {
          const value:Uint8Array = readerObject.value;
          const step: number = value.length;
          internalLog("Read bytes: " + step);
          // console.log(value);
          buffer.set(value, internalOffset);
          internalOffset += step;
          internalBytesRead += step;
          callbackCounter += 1;
        //  return callReadAndPromise();
        // } else {
          internalLog("No more data, total callbacks: " + callbackCounter);
          internalLog("Returning bytes : " + internalBytesRead);
          resolve({
            bytesRead: internalBytesRead,
            buffer: buffer
          });
        }
      });
    }
    
    if (this.boundedPort.readable) {
      if (this.classReader === undefined) {
        this.classReader = this.boundedPort.readable.getReader();
      }
      return basePromise.then(ok => {
        return new Promise((resolve, reject) => {
          if (this.isOpen) {
            // internalLog("WTF??");
            callReadAndPromise().then(resolve, reject);
          } else {        
            reject({
              canceled:true
            });
          }
        });
      });
    } else {
      return basePromise.then(ok => {
        return new Promise(reject => {
          reject({
            bytesRead: 0,
            buffer: buffer
          });
        });
      });
    }
  }

  update(options: { baudRate: number }): Promise<void> {
    internalLog("Update called");
    return super.update(options).then(ok => {
      return new Promise(reject => {
        reject(null);
      });
    });
  }

  set(options: SerialPort.SetOptions): Promise<void> {
    internalLog("Set called");
    return super.set(options).then(ok => {
      return this.boundedPort.setSignals({
        dataTerminalReady: options.dtr,
        requestToSend: options.rts,
        break: options.brk
      });
    });
  }

  // CTS, DSR, DCD
  get(): Promise<{
    cts: boolean;
    dsr: boolean;
    dcd: boolean;
    }> {
    internalLog("Get called");
    return super.get().then(ok => {
      this.boundedPort.getSignals().then(sis => {
        return new Promise((resolve, reject) => {
          if (sis) {
            resolve({
              cts: sis.clearToSend,
              dsr: sis.dataSetReady,
              dcd: sis.dataCarrierDetect
            });
          } else {
            reject(null);
          }
        });
      });
    });
  }

  getBaudRate(): Promise<number> {
    return new Promise(resolve => {
      resolve(this.actualBaudRate);
    })
  }

  flush(): Promise<void> {
    internalLog("Flush called");
    return super.flush().then(ok => {
      return new Promise(reject => {
        internalLog("Flush not supported");
        reject(null);
      });
    });
  }

  drain(): Promise<void> {
    internalLog("Drain called");
    return super.drain().then(ok => {
      return this.boundedPort.writable.close();
    });
  }
}


webpack.config.js
const path = require('path');
const { CleanWebpackPlugin } = require('clean-webpack-plugin');

const config = {
  target: 'web',
  entry: {
    index: './src/wsa-sp-bind.ts',
  },
  output: {
    path: path.resolve(__dirname, './dist'),
    filename: 'wsa-sp-bind.js',
    library: 'WSABinding',
    libraryTarget: 'umd',
    globalObject: 'this',
    umdNamedDefine: true,
  },
  watchOptions: {
    aggregateTimeout: 600,
    ignored: /node_modules/,
  },
  plugins: [
    new CleanWebpackPlugin({
      cleanStaleWebpackAssets: false,
      cleanOnceBeforeBuildPatterns: [path.resolve(__dirname, './dist')],
    }),
  ],
  module: {
    rules: [
      {
        test: /\.ts(x?)$/,
        exclude: [/node_modules/, /test/],
        use: [
          {
            loader: 'babel-loader',
          },
          {
            loader: 'ts-loader',
          },
        ],
      },
    ],
  },
  resolve: {
    extensions: ['.tsx', '.ts', '.js'],
  },
};

module.exports = (env, argv) => {
  if (argv.mode === 'development') {
    // * add some development rules here
  } else if (argv.mode === 'production') {
    // * add some prod rules here
  } else {
    throw new Error('Specify env');
  }

  return config;
};


src/wsa-sp-bind.ts
import AbstractBinding from '@serialport/binding-abstract';
import SerialPort, { PortInfo } from 'serialport';

declare global {
  interface Navigator {
    readonly serial: unknown;
  }
}

export default class WSABinding extends AbstractBinding {
  private static internalBufferSize = 16384;
  private static internalNavigatorSerial: any = navigator.serial;
  private static backendAvailable: boolean = navigator.serial !== undefined;
  private static internalBasePortsList: [...any] = [];

  private storedOpenOptions: SerialPort.OpenOptions = {
    baudRate: 115200,
    dataBits: 8,
    stopBits: 1,
    parity: 'none',
    rtscts: false,
  };
  private boundedPort: any | undefined = undefined;
  private actualBaudRate = -1;
  private classReader: ReadableStreamDefaultReader | undefined = undefined;
  private pendingRead: any | undefined = undefined;

  public static portFilters:
    | [{ usbVendorId: number; usbProductId?: number }]
    | undefined = undefined;
  public static debug = false;
  public isOpen = false;

  static list(): Promise<SerialPort.PortInfo[]> {
    return new Promise((resolve, reject) => {
      if (WSABinding.backendAvailable) {
        WSABinding.updatePortsList().then(() => {
          resolve(
            WSABinding.internalBasePortsList.map(
              WSABinding.mapBasePortToPortInfo,
            ),
          );
        });
      } else {
        reject();
      }
    });
  }

  constructor(opt: SerialPort.OpenOptions) {
    super(opt);
    Object.assign(this.storedOpenOptions, opt);
  }

  private static internalLog(text: string) {
    if (WSABinding.debug)
      console.log('%c WSA Log: ' + text, 'background: #222; color: #bada55');
  }

  private static dec2hex16bitWithPad(i: number): string {
    return (i + 0x10000).toString(16).substr(-4).toUpperCase();
  }

  private static mapBasePortToPath(sourceSerialPort: any): string {
    const sourceInfo: any = sourceSerialPort.getInfo();
    return (
      'wsa://' +
      WSABinding.dec2hex16bitWithPad(sourceInfo.usbVendorId) +
      '-' +
      WSABinding.dec2hex16bitWithPad(sourceInfo.usbProductId)
    );
  }

  private static mapBasePortToPortInfo(sourceSerialPort: any): PortInfo {
    const sourceInfo: any = sourceSerialPort.getInfo();
    return {
      path: WSABinding.mapBasePortToPath(sourceSerialPort),
      productId: sourceInfo.usbProductId,
      vendorId: sourceInfo.usbVendorId,
    };
  }

  private static updatePortsList(): Promise<void> {
    return WSABinding.internalNavigatorSerial.getPorts().then((ports) => {
      if (ports.length > 0) {
        return new Promise((resolve) => {
          WSABinding.internalBasePortsList = ports;
          resolve(null);
        });
      } else {
        if (WSABinding.internalBasePortsList.length == 0) {
          return WSABinding.internalNavigatorSerial
            .requestPort({ filters: WSABinding.portFilters })
            .then((port) => {
              WSABinding.internalBasePortsList = [port];
              return new Promise((resolve) => {
                resolve(null);
              });
            });
        } else {
          return new Promise((resolve) => {
            resolve(null);
          });
        }
      }
    });
  }

  private static mapOpenOptions(input: SerialPort.OpenOptions): any {
    return {
      baudRate: input.baudRate,
      dataBits: input.dataBits,
      stopBits: input.stopBits,
      parity: input.parity,
      bufferSize: WSABinding.internalBufferSize,
      flowControl: input.rtscts ? 'hardware' : 'none',
    };
  }

  open(path: string, options: SerialPort.OpenOptions): Promise<void> {
    WSABinding.internalLog('open called, path: ' + path);
    this.classReader = undefined;
    return super.open(path, options).then(() => {
      return WSABinding.updatePortsList().then(() => {
        for (const port of WSABinding.internalBasePortsList) {
          const mappedName = WSABinding.mapBasePortToPath(port);
          WSABinding.internalLog('Checking ' + mappedName);
          if (mappedName == path || path == 'wsa://default') {
            this.boundedPort = port;
            const localOptions = this.storedOpenOptions;
            Object.assign(localOptions, options);
            return this.boundedPort
              .open(WSABinding.mapOpenOptions(localOptions))
              .then(
                () => {
                  WSABinding.internalLog('Open successful');
                  this.isOpen = true;
                  return new Promise((resolve) => {
                    WSABinding.internalLog('Open resolved');
                    resolve(null);
                  });
                },
                (failAny) => {
                  this.isOpen = false;
                  WSABinding.internalLog('Open failed');
                  return new Promise((reject) => {
                    WSABinding.internalLog('Open rejected with ' + failAny);
                    reject(null);
                  });
                },
              );
          }
        }
      });
    });
  }

  private closePromise(): Promise<void> {
    return this.boundedPort.close().then(() => {
      return new Promise((resolve) => {
        this.isOpen = false;
        resolve(null);
      });
    });
  }

  close(): Promise<void> {
    WSABinding.internalLog('close called');
    return super.close().then(() => {
      if (this.classReader) {
        this.classReader.cancel().then(() => {
          this.classReader.releaseLock();
          return this.closePromise();
        });
      } else {
        return this.closePromise();
      }
    });
  }

  write(buffer: Buffer): Promise<void> {
    WSABinding.internalLog('write called');
    const mWriter = this.boundedPort.writable.getWriter();
    return super.write(buffer).then(() => {
      return mWriter.write(buffer).then(() => {
        return new Promise((resolve) => {
          WSABinding.internalLog('Write system execution completed');
          mWriter.releaseLock();
          resolve(null);
        });
      });
    });
  }

  private registerInternalRead(): Promise<ReadableStreamReadResult<any>> {
    const internalRead = this.classReader.read();
    this.pendingRead = internalRead;
    return internalRead;
  }

  private clearInternalRead() {
    this.pendingRead = undefined;
  }

  read(
    buffer: Buffer,
    offset: number,
    length: number,
  ): Promise<{ bytesRead: number; buffer: Buffer }> {
    let internalOffset: number = offset;
    let internalBytesRead = 0;
    let callbackCounter = 0;
    const currentRef: WSABinding = this;

    WSABinding.internalLog('read called, bytes expecting: ' + length);
    const basePromise: Promise<any> = super.read(buffer, offset, length);

    function callReadAndPromise(): Promise<{
      bytesRead: number;
      buffer: Buffer;
    }> {
      WSABinding.internalLog(
        'Read system execution initiated ' + callbackCounter,
      );
      return currentRef.registerInternalRead().then((readStruct) => {
        WSABinding.internalLog(
          'Read system execution completed ' + callbackCounter,
        );
        currentRef.clearInternalRead();
        return internalPromise(readStruct);
      });
    }

    function internalPromise(readerObject: {
      value?: any;
      done: boolean;
    }): Promise<{ bytesRead: number; buffer: Buffer }> {
      return new Promise<{ bytesRead: number; buffer: Buffer }>((resolve) => {
        WSABinding.internalLog('Updating result buffer ' + callbackCounter);
        // const done = readerObject.done;
        if (readerObject.value) {
          // we ignore the "done" flag, as it seems never to
          // reach the point of being set, even when there is
          // no more data to read, the Web Serial Api will just block
          // if (!done) {
          const value: Uint8Array = readerObject.value;
          const step: number = value.length;
          WSABinding.internalLog('Read bytes: ' + step);
          // console.log(value);
          buffer.set(value, internalOffset);
          internalOffset += step;
          internalBytesRead += step;
          callbackCounter += 1;
          //  return callReadAndPromise();
          // } else {
          WSABinding.internalLog(
            'No more data, total callbacks: ' + callbackCounter,
          );
          WSABinding.internalLog('Returning bytes : ' + internalBytesRead);
          resolve({
            bytesRead: internalBytesRead,
            buffer: buffer,
          });
        } else {
          resolve({
            bytesRead: 0,
            buffer: buffer,
          });
        }
      });
    }

    if (this.boundedPort.readable) {
      if (this.classReader === undefined) {
        this.classReader = this.boundedPort.readable.getReader();
      }
      return basePromise.then(() => {
        return new Promise((resolve, reject) => {
          if (this.isOpen) {
            // WSABinding.internalLog("WTF??");
            callReadAndPromise().then(resolve, reject);
          } else {
            reject({
              canceled: true,
            });
          }
        });
      });
    } else {
      return basePromise.then(() => {
        return new Promise((reject) => {
          reject({
            bytesRead: 0,
            buffer: buffer,
          });
        });
      });
    }
  }

  update(options: { baudRate: number }): Promise<void> {
    WSABinding.internalLog('update called');
    return super.update(options).then(() => {
      return new Promise((reject) => {
        WSABinding.internalLog('update not supported');
        reject(null);
      });
    });
  }

  set(options: SerialPort.SetOptions): Promise<void> {
    WSABinding.internalLog('set called');
    return super.set(options).then(() => {
      return this.boundedPort.setSignals({
        dataTerminalReady: options.dtr,
        requestToSend: options.rts,
        break: options.brk,
      });
    });
  }

  // CTS, DSR, DCD
  get(): Promise<{
    cts: boolean;
    dsr: boolean;
    dcd: boolean;
  }> {
    WSABinding.internalLog('get called');
    return super.get().then(() => {
      this.boundedPort.getSignals().then((singalsInfoStructure) => {
        return new Promise((resolve, reject) => {
          if (singalsInfoStructure) {
            resolve({
              cts: singalsInfoStructure.clearToSend,
              dsr: singalsInfoStructure.dataSetReady,
              dcd: singalsInfoStructure.dataCarrierDetect,
            });
          } else {
            reject(null);
          }
        });
      });
    });
  }

  getBaudRate(): Promise<number> {
    WSABinding.internalLog('getBaudRate called');
    return new Promise((resolve) => {
      resolve(this.actualBaudRate);
    });
  }

  flush(): Promise<void> {
    WSABinding.internalLog('flush called');
    return super.flush().then(() => {
      return new Promise((reject) => {
        WSABinding.internalLog('flush not supported');
        reject(null);
      });
    });
  }

  drain(): Promise<void> {
    WSABinding.internalLog('drain called');
    return super.drain().then(() => {
      return new Promise((reject) => {
        WSABinding.internalLog('drain not supported');
        reject(null);
      });
    });
  }
}
